### ISSUE TITLE
getContent on StreamedResponse shouldn't return false

### ISSUE THREAD

**Issue by jpswade** [symfony/symfony#44415]
### Symfony version(s) affected

all

### Description

When trying to debug an issue with the StreamedResponse, and trying to getContent, you get a false, however, you would expect an exception, rather than false, as you get an exception when you try to set content, which makes it clear.

If you can't getContent on an object, it should throw an exception, because it's exceptional circumstance rather than returning false.

Returning false gives the impression that you can set content, it gives the impression that the content is false, when actually it's not possible to set the content, and it should throw an exception because you can't get content on a StreamedResponse.

I'm raising this as an issue rather than a PR as I'm interested to hear why it might be useful for it to return a false.

### How to reproduce

$response = new StreamedResponse();
$response->getContent();

### Possible Solution

Throw an exception.

**Comment by nicolas-grekas** [comment]
This is by design. The return type of the method is documented as `string|false` on `Response::getcontent()` precisely to cover this need for streamed responses.

**Comment by jpswade** [comment]
> This is by design. The return type of the method is documented as `string|false` on `Response::getcontent()` precisely to cover this need for streamed responses.

Yes I'm aware it's by design and that the document says that, however, the issue is that it's a poor design and implementation, hence raising this issue.

As `getContent` is not implemented, precisely for this reason, it should say so, rather than failing silently.

**Comment by nicolas-grekas** [comment]
I'm afraid changing this would be a too costly change for what it's worth (taking BC/FC into account of course.)

**Comment by jpswade** [comment]
> I'm afraid changing this would be a too costly change for what it's worth (taking BC/FC into account of course.)

I'm not sure I understand how you've calculated that cost, but it doesn't seem costly to me as the method is unusable.

The interface shouldn't need to change as the return type wouldn't change.

Only it would instead now, correctly throw an exception in an exceptional circumstance (ie: when the method is not implemented).

In this situation, `getContent` always returns false, so adding an exception (that can be caught as necessary), seems like a reasonable way to communicate that the content is not false, but that the method is actually not implemented and cannot be used.

The false would then be reserved for situations when the content is genuinely false, rather than than exceptional situations where the method isn't implemented.

**Comment by nicolas-grekas** [comment]
But there are no situations where returning false is genuine. Throwing an exception will break existing code that deal with false to detect when the body is streamed instead. We would require a way to tell these existing codes that the behavior is going to change in V7. But since this API dates from 10 years back, the impact is likely high. I don't have numbers and won't be able to have any, but taking a cautious approach is required on such topics.
Changing should provide a significant value compared to the cost. Yes, the real cost is unknown. But the real value isn't either to me. By experience, I wouldn't change this.

**Comment by jpswade** [comment]
If the content can never be genuinely false, then that's more the reason to get rid of it.

I think you'd be hard pushed to find an example in the wild of the implementation you mention.

I understand what it means to maintain backwards compatibility, I worked on backwards compatibility libraries when we moved from PHP4 to PHP5, however it didn't stop the language from progressing, I don't see why it should be any different here.

Ten years ago, when this was implemented, PHP was still on 5.4, much has moved on since then, and that's reflected in the fact that returning false here is an outdated implementation and this should be updated to reflect that.

You're right though, the value to make this change isn't there for those more experienced with the framework, only those coming in fresh. It's definitely much easier to keep things the same, even if it's a bad user experience.

**Comment by wouterj** [comment]
Backwards compatibility means we have to stay backwards compatible. Immediately throwing an exception isn't backwards compatible. If you have any suggestion to warn users (using deprecation notices) that they should change

```php
if ($contents = $response->getContent()) {
}
```

to

```php
try {
    $contents = $response->getContent();
} catch (\BadMethodCallException $e) {
}
```

feel free to create a PR (btw, looking at the code, I'm not sure if the exception really is a better API?).

However, it's clear that at least 3 core team members are not going to spent their evenings thinking about a solution here :)
